wemake_python_styleguide/logic/returns.py

```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+
+import ast
+from collections import defaultdict
+from typing import Dict, List, Tuple
+
+NamesAndReturns = Tuple[
+    Dict[str, List[ast.Name]],
+    Dict[str, ast.Return],
+]
+
+
+def get_assign_node_variables(node: List[ast.AST]) -> List[str]:
+    """Gets the List of assign ids of the nodes."""
+    assign = []
+    for sub_node in node:
+        if isinstance(sub_node, ast.Assign):
+            if isinstance(sub_node.targets[0], ast.Name):
+                assign.append(sub_node.targets[0].id)
+        if isinstance(sub_node, ast.AnnAssign):
+            if isinstance(sub_node.target, ast.Name):
+                assign.append(sub_node.target.id)
+    return assign
+
+
+def get_name_nodes_variable(node: List[ast.AST]) -> Dict[str, List[ast.Name]]:
+    """Gets the List of names of the nodes."""
+    names: Dict[str, List[ast.Name]] = defaultdict(list)
+    for sub_node in node:
+        if isinstance(sub_node, ast.Name):
+            if isinstance(sub_node.ctx, ast.Load):
+                names[sub_node.id].append(sub_node)
+        if isinstance(sub_node, ast.AugAssign):
+            if isinstance(sub_node.target, ast.Name):
+                variable_name = sub_node.target.id
+                names[variable_name].append(sub_node.target)
+    return names
+
+
+def get_return_node_variables(node: List[ast.AST]) -> NamesAndReturns:
+    """Gets the return values and the return sub nodes in the code."""
+    returns: Dict[str, List[ast.Name]] = defaultdict(list)
+    return_sub_nodes: Dict[str, ast.Return] = defaultdict()
+    for sub_node in node:
+        if isinstance(sub_node, ast.Return):
+            if isinstance(sub_node.value, ast.Name):
```

Comment:
For the correct_example_N we are not collecting the previous returns where I expected them to be collected in return_sub_nodes and so we are running into issues here. The correct_example_N1 is passing the tests. How should we collect those returns here? Example the returns of type ast.Attribute, and should we be collecting them here?
@sobolevn

Reply:
Sorry, I have missed your question. Can you try to use `ast.walk(subnode)` there? I guess that happens because they are not direct children of the give node, but are lower down the tree.

Reply:
You can try some other task if you are stuck with one. And pass this one to me.

Reply:
Sure